DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces_overview/host-pipes/src/vector_add.cpp

```diff
@@ -0,0 +1,116 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>  // sycl.hpp comes first to prevent compiler crash
+
+#define VECT_SIZE 256
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class IDSimpleVAddPipes;
+
+using PipeProps = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::ready_latency<0>));
+
+class IDPipeA;
+using InputPipeA =
+    sycl::ext::intel::experimental::pipe<IDPipeA, int, 0, PipeProps>;
+
+class IDPipeB;
+using InputPipeB =
+    sycl::ext::intel::experimental::pipe<IDPipeB, int, 0, PipeProps>;
+
+class IDPipeC;
+using OutputPipeC =
+    sycl::ext::intel::experimental::pipe<IDPipeC, int, 0, PipeProps>;
+
+
+struct SimpleVAddKernelPipes {
```

Comment:
```suggestion
template <typename InputPipeA,
	          typename InputPipeB,
	          typename OutputPipeC
                > 
struct SimpleVAddKernelPipes {
```

Reply:
@chanjunc-intel  Do you find it easier to read code when the pipes are templated like this? I find templating code like this makes messy debug messages.

Reply:
I agree with you on this. The code will be harder to read. But I guess templating do provide more flexibility in reusing in larger design. I'm suggesting this here as the kernel call is using this template-based style.